src/libostree/ostree-sysroot-cleanup.c

```diff
@@ -216,6 +216,15 @@ cleanup_other_bootversions (OstreeSysroot       *self,
   const int cleanup_subbootversion = self->subbootversion == 0 ? 1 : 0;
   /* Reusable buffer for path */
   g_autoptr(GString) buf = g_string_new ("");
+  struct stat stbuf;
+
+  if ((glnx_fstatat_allow_noent (self->sysroot_fd, "boot/loader", &stbuf, AT_SYMLINK_NOFOLLOW, error))
+      && (!S_ISLNK (stbuf.st_mode)))
+    {
+      g_string_truncate (buf, 0); g_string_append_printf (buf, "boot/loader.%d", self->bootversion);
+      if (!glnx_shutil_rm_rf_at (self->sysroot_fd, buf->str, cancellable, error))
+        return FALSE;
+    }
```

Comment:
Is this logic even right? In `swap_bootloader`, `loader.<newversion>` is exchanged with `loader`. That means that the current bootloader directory is now named `loader.<newversion>`. What's even at `loader.<currentversion>`? This is also in conflict with the code below that cleans up `loader.<currentversion>`. So, after this change, there's no more `loader.*` bootloader directories left. Is that correct?

Reply:
Right. Only `loader` directory is left after deployment is finished successfully. We could just use `loader.tmp` and `loader`. But if I remember correctly, we need to support older deployments that are still using symlinks. For that reason we still to need to keep this 0/1 version names. Also maintaining a patch for long term I cannot really afford changing too much the logic.

Reply:
Reading the code, I do not think the support for older deployment is working. I will try to add tests for that.